### ISSUE TITLE
Use openssl::sha256

### ISSUE THREAD

**Issue by wilkox** [wilkox/deidentifyr#2]

**Comment by jonocarroll** [comment]
Did you find this actually improved performance? 

I have built an internal package at work which performs this exact task (same name, no less; snap!). I have been using `digest` but it has issues, not least of which is that it's not vectorized over the inputs. `openssl::sha256()` is however, so you can do without the `sapply()` for a neat 10x boost

``` r
microbenchmark::microbenchmark(
  digest  = sapply(as.character(1:1e3), function(x) digest::digest(x, algo = "sha256", serialize = FALSE)),
  sapply  = sapply(as.character(1:1e3), openssl::sha256, USE.NAMES = FALSE),
  direct = openssl::sha256(as.character(1:1e3))
)
#> Unit: milliseconds
#>    expr       min        lq      mean    median        uq       max neval cld
#>  digest 27.527302 32.422624 35.568593 35.518777 38.051679 61.583824   100  b
#>  sapply 39.550945 44.961160 50.926576 49.360425 53.146786 91.503245   100   c
#>  direct  3.080855  3.458596  3.744925  3.718039  4.003392  4.942569   100 a

identical(
  unclass(openssl::sha256(as.character(1:5))),
  sapply(as.character(1:5), openssl::sha256, USE.NAMES = FALSE)
)
#> [1] TRUE
```

Another (now obsolete) advantage which you may not have identified is that `digest` serializes objects by default before hashing, which means the resulting hash is not comparable via other means. I had to cross-check mine against a python implementation to be sure, but a comparison with `sha256()` shows it's not serializing

``` r
identical(
  unclass(openssl::sha256(as.character(1:5))),
  sapply(as.character(1:5), function(x) digest::digest(x, algo = "sha256", serialize = FALSE), USE.NAMES = FALSE)
)
#> [1] TRUE

hl <- reticulate::import("hashlib", convert = FALSE)
identical(
  unclass(openssl::sha256(as.character(1:5))),
  sapply(as.character(1:5), function(x) as.character(hl$sha256(x)$hexdigest()), USE.NAMES = FALSE)
)
#> [1] TRUE
```

It seems you're aware of the increased risk of collision with a truncated hash; do you have any thoughts about how to address that in the scenario where a user encounters collisions? It seems at the moment you throw a warning, which I think is about as far as I managed too. If your input set has more than one `NA` value the warning will be tripped.

**Comment by wilkox** [comment]
I switched to `openssl::sha256()` for the vectorisation boost, though to be honest I didn't actually measure and compare performance beyond 'this seems faster'. Thanks for pointing out that `digest` was serialising the inputs, I hadn't even considered that.

I also couldn't come up with a good solution for genuine hash collisions. The best idea I have so far is to keep the entire 64-character hashes, but in an S3 class with `format`, `print` etc. methods that would only print the first ten or so characters to avoid filling the screen with garbage. The downside is it would be a lot harder to subset by ID. For now, ten characters provides pretty good odds so I'm not too worried about collisions.

**Comment by jonocarroll** [comment]
That's somewhat odd, since you don't take advantage of the vectorization, you're still passing the values through `sapply()` (hence the slowdown in my above comparison)

https://github.com/wilkox/deidentifyr/blob/d1d708bc3e4e32bf54ebf1503f1e6999b6493296/R/deidentify.R#L58

i.e. replace with `hashes <- openssl::sha256(input)`.

A `print` method would be a great solution; I think I'll do the same in my implementation. I have no intentions to subset by hash so this solves the problem for me. Extracting by substring would just mean a custom `[` method as well (or an `extract()` function).

I ran some tests on our large clusters and scanned the entire likely parameter space of potential inputs (~ tens of billions). With 10 characters I got some collisions. I managed to drop the likelihood significantly by using 12 characters. I ended up passing the truncation length in as a parameter for now. Our plan is to map out the space relevant to a single trial for collisions before we hash anything.
